engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala

```diff
@@ -181,15 +180,38 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
     val actionThreadPoolSize = databaseConfig.getIntOr("actionThreadPoolSize", dbNumThreads) min dbMaximumPoolSize
     Executors.newFixedThreadPool(actionThreadPoolSize)
   }
+
   private val actionExecutionContext: ExecutionContext = ExecutionContext.fromExecutor(
-    actionThreadPool, database.executor.executionContext.reportFailure)
+    actionThreadPool, database.executor.executionContext.reportFailure
+  )
+
+  private lazy val useSlickUpserts = dataAccess.driver.capabilities.contains(JdbcProfile.capabilities.insertOrUpdate)
 
   override def close(): Unit = {
     actionThreadPool.shutdown()
     database.close()
   }
 
-  // Run action with an outer transaction
+  private def runTransactionWithRollbackRetry[R](action: DBIOAction[R, _ <: NoStream, _ <: Effect], actorSystem: ActorSystem): Future[R] = {
```

Comment:
looks like this can be `action: DBIO[R]`, at least when `runTransaction` below is changed similarly